### DISCUSSION TITLE
Custom TeX templates in zettlr 2.0beta3 ?

### DISCUSSION THREAD

**Discussion by BP22190** [Zettlr/Zettlr#2371] [Q&A]
Hello,

I'm trying zettlr 2.0beta3, and don't find how to use/specify a TeX custom template... 

Any way ?

Best regards,

**Comment by hildebrand-institute** [comment]
So far this is my way to go:

1) File-->Preferences-->Assets Manager

2) tab "Exporting" and then on the left panel-->PDF

3) Wherever you want (I've put it at the end of the default file) write "defaults:" followed by the absolute path to a second YAML cofig file. The idea is that of using a secondary config YAML file for each of your final pdf documents. This is e.g. what I've written within the default file for finally producing one of my last paper: 

`defaults: /run/media/alex/DATA/zettlr_defaults_files/practical_relativism.yaml`

4) Within the second YAML file (in this case: "partical_relativism.yaml") you can write all the specific cofigurations that concern the particular document your are going to produce through pandoc to LaTeX/XeLaTeX. This is the content of mine:

```
template: path_to_the_tex_template
bibliography:
- path_to_the_bib_file
```

In the pandoc manual section "Default file" you can find all the other valid configuration options that can be used. Hope this is useful. Note that I'm not an expert, as I began experimenting with Zettlr 2.0.0 only in the last weekend.

Cheers
Alex

**Reply by nathanlesage** [reply]
This sounds like a little bit too complicated, since you can also specify stuff in the YAML frontmatter (especially bibliography, this should be inside the YAML frontmatter).

Templates, however, can only be specified in the defaults file, but instead of pointing to a different defaults file, I would suggest you add a property `template: /path/to/template.tex`. Unfortunately you have to specify the template globally, BUT you can simply comment out the template, and thus easily use several templates as you like:

```yaml
# ZETTLR DEFAULTS FILE
# ====================
# Conversion: Markdown --> PDF (requires TeX suite installed)
# More info: https://pandoc.org/MANUAL.html
reader: markdown
writer: pdf
template: my-template.tex # Currently used
# template: cv-template.tex
# template: special-template.tex
# template: yet-another-template.tex
self-contained: true
# ... snip ...
```

This way you can just comment out whichever template you want to use. Still a little bit of a hassle, but the Pandoc team is aware of this limitation, so they'll probably add the possibility to specify a template in the YAML frontmatter.

**Reply by hildebrand-institute** [reply]
After a better reflection. Since what I can write inside a default file can be written also inside a YAML front matter, you are totally right, there's no necessity of the second YAML file...

**Reply by nathanlesage** [reply]
Attention: Not everything that works in a defaults file can be added to the frontmatter. The `template`-property, for instance, doesn't work! Unfortunately, there's no list for that, so it requires a little bit of trial and error to find out what works and what not.

**Comment by BP22190** [comment]
Hello,

Thanks for your replies !

Another question about that : is there a way to customize a Pandoc command ? (i.e. a way for pandoc options)

Best regards,

**Reply by hildebrand-institute** [reply]
I'm not sure to understand correctly your question. Previous versions of Zettlr were based upon customized pandoc commands. Now it seems that what previously was obtained through customized commands can be obtained through the default file togheter with YAML frontmatters..., if I'm not wrong.

**Reply by nathanlesage** [reply]
This: The default files allow you to do everything you could also do via the command-line (but even more, which is why we switched)

**Reply by mj-designthinks** [reply]
One reason to customize the Pandoc command is to enable various [extensions](https://pandoc.org/MANUAL.html#extensions) to markdown. For example, I primarily use these extensions in my workflow:
`
pandoc -f markdown+smart+definition_lists+mmd_title_block+strikeout+bracketed_spans+fenced_divs 
`

I was able to accomplish this in the previous versions of Zettlr, but am not sure how to accomplish this in the defaults.

**Reply by BP22190** [reply]
_"I was able to accomplish this in the previous versions of Zettlr, but am not sure how to accomplish this in the defaults."_

Yes... me too, I can't do with this version what I did with the previous one, it's a big problem for me. To the point of wondering if I continue to use Zettlr if we fear that there will not be backward functional compatibility with new versions ...

**Reply by nathanlesage** [reply]
This is of course also possible with defaults files! You really don't need to provide any commands, that's the beauty of these. As you will see (arguably somewhat hidden!) in [the documentation on defaults files](https://pandoc.org/MANUAL.html#default-files), they put this in the reader section:

```yaml
from: markdown+emoji
# reader: may be used instead of from:
to: html5
# writer: may be used instead of to:
```

As you can see: the `-f` command line argument is exactly the same as the `from` (or `reader`) property and works exactly the same. So in the case of @mj-designthinks you could use this example:

```yaml
from: markdown+smart+definition_lists+mmd_title_block+strikeout+bracketed_spans+fenced_divs
to: html5
```

**Reply by mj-designthinks** [reply]
Yes, I discovered that markdown+extensions works in the defaults file. I was hesitant to do so because the Zettlr docs warn against changing the reader or the writer properties.

<img width="751" alt="image" src="https://user-images.githubusercontent.com/37943655/131578094-6952bee7-6cb9-4daf-8fac-9fbfceb5d246.png">

I guess the exception would be adding extensions to the markdown reader, separated by `+`. I don't know if any of the other readers have extensions.

**Reply by nathanlesage** [reply]
That is true! We should relax this warning.

**Comment by vsanchez** [comment]
It would be nice to add the option to have a local project default file, that may make everything cleaner and more modular and (I think) not too hard to implement:

In the project Properties popup:

1. In the "Formats" tab add: "Project defaults" as an option (besides the Zettlr defaults)
2. In the "Files" tab add: Project defaults: (a file selector)

If "Project defaults" is selected and there is a file defined in "Project defaults" then pass that option to Pandoc.

**Reply by nathanlesage** [reply]
Could be a potentially good idea, feel free to open an issue!
